Complete release protocol (#601)
rl-institut/multi-vector-simulator: CONTRIBUTING.md

## Release protocol

> ### Before the release

Reviewer: ```suggestion
### Before the release on pypi.org
```

1. Create a release branch by branching off from `dev`
    ```bash
    git checkout -b release/vX.Y.Z dev
    ```
    For meaning of X, Y and Z version numbers, please refer to this [semantic versioning guidelines](https://semver.org/spec/v2.0.0.html).
2. In your release branch, update the version number in [`src/multi-vector-simulator/version.py`](https://github.com/rl-institut/multi-vector-simulator/blob/dev/src/multi_vector_simulator/version.py) (commit message: "Bump version number") and

> in [`CHANGELOG.md`](https://github.com/rl-institut/multi-vector-simulator/blob/dev/CHANGELOG.md) in the in 1. indicated format.

Reviewer: ```suggestion
in [`CHANGELOG.md`](https://github.com/rl-institut/multi-vector-simulator/blob/dev/CHANGELOG.md) in the in the format indicated under 1.
```

3. Add the date of release in the `Changelog.md` in [ISO format](https://xkcd.com/1179/).
4. After pushing these changes, create a pull request from `release/vX.Y.Z` towards `master` and merge it into `master`.

> ### The actual release

Reviewer: ```suggestion
### The actual release on pypi.org
```

1. Open a working python3 virtual environment and make sure you have the latest versions of setuptools and wheel installed:

> `python3 -m pip install --user --upgrade setuptools wheel twine`

Reviewer: ```suggestion
`python3 -m pip install --upgrade setuptools wheel twine`
```
--user option gave me an error code this morning

> 2. Delete `build/` and `dist/` directory to avoid errors.

Reviewer: ```suggestion
2. Delete `build/` and `dist/` directories (from previous release) to avoid errors.
```

`python3 setup.py sdist bdist_wheel`
4. Check the package with twine: `python3 -m twine check dist/*` If errors occur, fix them before the release or postpone.
5. Create a [release tag](https://github.com/rl-institut/multi-vector-simulator/releases) on github.
Please choose `master` as target and use `vX.Y.Z` as tag version. In the description field simply copy-paste the content of the `CHANGELOG`descriptions for this release and you're done!
For help look into the [github release description](https://help.github.com/en/github/administering-a-repository/creating-releases).
6. Pull `master` to your local repository and check if the new tag is there. (If you do not see the tag type `git fetch`.)
7. Test your package:
    1. Test the upload on test.pypi.org
        ` twine upload --repository testpypi dist/*`
    2. Test the installation: `pip install -i https://test.pypi.org/simple/ multi-vector-simulator --no-deps`
    3. Then open a python console and try
        `>>>from multi_vector_simulator import cli`
8. If everything works as expected you can now upload the package version to PyPI
    1. Check the credentials of our pypi@rl-institut.de account on https://pypi.org.

>     1. Type `twine upload dist/*`

Reviewer: ```suggestion
    2. Type `twine upload dist/*`
```

Author: Funny fact: This would still be shown as "1., 2. etc." on readthedocs no matter which number is written. But still we can get it right in the md-file.

>     2. Enter `__token__` for username and your pypi token for password. 

Reviewer: ```suggestion
    3. Enter `__token__` for username and your pypi token for password. 
```

## Release protocol

### Before the release

1. Create a release branch by branching off from `dev`
    ```bash
    git checkout -b release/vX.Y.Z dev
    ```
    For meaning of X, Y and Z version numbers, please refer to this [semantic versioning guidelines](https://semver.org/spec/v2.0.0.html).
2. In your release branch, update the version number in [`src/multi-vector-simulator/version.py`](https://github.com/rl-institut/multi-vector-simulator/blob/dev/src/multi_vector_simulator/version.py) (commit message: "Bump version number") and
in [`CHANGELOG.md`](https://github.com/rl-institut/multi-vector-simulator/blob/dev/CHANGELOG.md) in the in the format indicated under 1.

> Attention: Please work with release candidates (add `rc1` to the `version_num`, for example `v0.5.0rc1`) before the actual release, as a release with a specific version number can only be done once.

Reviewer 2: Mh, okay. Will need to see how this works, as I do not understand how I would have to do a release now.

Author: Could you specify what you don't understand? Then we can explain better here. Is it the release candidates?

> 3. Add the date of release in the `Changelog.md` in [ISO format](https://xkcd.com/1179/).

Reviewer 2: Maybe "replace the changelog header `[Unreleased]` with the number and date of the release `[Version] - YYYY-MM-DD`"?

> 4. After pushing these changes, create a pull request from `release/vX.Y.Z` towards `master` and merge it into `master`.

Reviewer 2: How does this work with the release candidates? Isnt it `release/vX.Y.Zrc1` then?

Reviewer: release candidate are only for pypi. @SabineHaas I think merging the two release protocols (release on master and release on pypi) into one might be confusing, maybe we need to specify which "release" we mean?

Author: Ah maybe I mixed something up here because I never used release candidates before.
Do you do the releases from the `rcX` branch? Cause then we need do describe the steps differently. 
I hope it won't get too confusing then. 
I don't really have time to work on this these days.. so if you wanna do a suggestion - go for it:)

Reviewer: My suggestion would have been to make two discting entries in the CONTRIBUTING, one for a release on master, and another one for release/upload on pypi. The rationale behind it, is that the release on master can be done by almost everyone, whereas the release on pypi can only be done by people with credentials

### The actual release
1. Open a working python3 virtual environment and make sure you have the latest versions of setuptools and wheel installed:
`python3 -m pip install --upgrade setuptools wheel twine`
2. Delete `build/` and `dist/` directories (from previous release) to avoid errors.
3. Make sure you are in `master` and the latest commit you want to release and create the package:
`python3 setup.py sdist bdist_wheel`
4. Check the package with twine: `python3 -m twine check dist/*` If errors occur, fix them before the release or postpone.
5. Create a [release tag](https://github.com/rl-institut/multi-vector-simulator/releases) on github.

> Please choose `master` as target and use `vX.Y.Z` as tag version. In the description field simply copy-paste the content of the `CHANGELOG`descriptions for this release and you're done!

Reviewer 2: Is the tag then overwritten if we have to do a `rc2`? Or should the tag be `vX.Y.Zrc1`?

    2. Test the installation: `pip install -i https://test.pypi.org/simple/ multi-vector-simulator --no-deps`
    3. Then open a python console and try
        `>>>from multi_vector_simulator import cli`
8. If everything works as expected you can now upload the package version to PyPI
    1. Check the credentials of our pypi@rl-institut.de account on https://pypi.org.
    2. Type `twine upload dist/*`
    3. Enter `__token__` for username and your pypi token for password. 
9. If your release candidate works well you can now do the actual release: start from step 2 in "Before the release" and remove `rcX` from [`src/multi-vector-simulator/version.py`](https://github.com/rl-institut/multi-vector-simulator/blob/dev/src/multi_vector_simulator/version.py).
If you notice errors in the uploaded package, fix them and bump up `rc1` to `rc2` until you don't see any more mistakes.

### After the release

1. Locally, merge `release/vX.Y.Z` into `dev` and push to the remote version of dev.
--> The idea is to avoid creating a merge commit in `dev` (because `master` would otherwise have two merge commits for this release once you merge the next release).

> 2. Set version for next release in [`src/multi-vector-simulator/version.py`](https://github.com/rl-institut/multi-vector-simulator/blob/dev/src/multi_vector_simulator/version.py) of the `dev`: for example `0.5.1dev`

Reviewer 2: I do not understand why I have to define a version here and why define the version number with suffix `dev`. Is it so that the version is printed in the reports and does not link mistakenly to the real release while the dev branch is used for the simulation?

Author: This is the procedure I learned in other projects. And yes, I think it's useful to show you the version you are working with is not the `0.5.1` in this case but a version that is developed from `0.5.1`.

> 3. Add the structure for a new `unreleased` version to the [`CHANGELOG.md`](https://github.com/rl-institut/multi-vector-simulator/blob/dev/CHANGELOG.md) in `dev`:

Reviewer 2: Push directly to dev?

Reviewer: This is one of the rare example where you want to push directly to dev after merging the release branch locally The idea is to avoid having merge commit of the same release twice in `master`: once in `master` when we release and at the next release (if we had a release commit in dev, this one will eventually end up into `master` at the next release)

Author: What do you mean by that? Adding a comment saying "push to dev"?

Author: > What do you mean by that? Adding a comment saying "push to dev" to make sure that is does not stay locally?
